### ISSUE TITLE
Why is the mdcache entry inserted into L1 by LRU, not MRU?

### ISSUE THREAD

**Issue by sepia-liu** [nfs-ganesha/nfs-ganesha#506]
In lru_reap_impl, always determine whether the first entry(head->next) of each lane is recyclable. It may appear that the entry to be reaped is the entry being created。

**Comment by dang** [comment]
The reason is that lots of entries are created but never used again (such as during readdir), and we don't want them to push the active entries out of the MRU end.  If they're actually used after being created, then they'll be immediately moved to the MRU end.

**Comment by sepia-liu** [comment]
Thank you,but if insert L1 by LRU, it will appear that the entry to be reaped is the entry being created，the entry cannot be recycled and can only to new a mdcache entry finally ( actually the lane has reclaimable entry).
When i create 40000 files, 100 threads，set threshold  Entries_HWMark = 5000;
The last entries_uesd = 39391, far beyond the set threshold. So I think it's flawed。 @dang

**Comment by dang** [comment]
Actually, Entries in L1 can be recycled.  If you look at lru_try_reap_entry(), you'll see that it reaps from L2 first, and then from L1 if L2 failed.  So any entry eligible for reaping can be recycled.  Eligible means that nothing currently has a refcount on that entry.

**Comment by sepia-liu** [comment]
Maybe what I want to ask is, when the entries_uesd is greater than the threshold value, the entry should be recycled from the already created entries. Why are so many new entries created(far beyond the set threshold)? 
So I suspect that the entry being created(ref = 2) in head->next, at the same time it also used to determine whether it is recyclable.Obviously unsatisfied and can only find next lane(so do other lanes).

**Comment by dang** [comment]
In most cases, the created entry will also be ref'd and used, so it won't be eligible to be recycled until it's done.  A common case when this would happen is readdir, although it shouldn't create thousands of them.  Which version of Ganesha are you using, and what readdir related config are you using?  It's possible there's a refcount leak (there have been some in the past) that's keeping entries from being recycled.

**Comment by sepia-liu** [comment]
I use ganesha 2.7.6 version, ganesha.conf as follows:
```
CACHEINODE
{
        Dir_Chunk = 0;
        NParts = 1;
        Cache_Size = 1;
        Entries_HWMark = 5000;
        FD_LWMark_Percent = 10;
}
EXPORT
{
        Export_Id = 1;
        Path = /nfs-test1;
        Pseudo = /nfs-test1;
        Access_Type = RW;
        Protocols = 3,4;
        Squash = No_Root_Squash;
        Filesystem_id = 666.1;
        Attr_Expiration_Time = 0;
        FSAL {
                Name = CEPH;
        }
}
```

I'm trying to change it to MRU in L1. The Entries_HWMark works.

**Comment by ffilz** [comment]
This one may bear more conversation before closing.

**Comment by ffilz** [comment]
Daniel, does our recent discussion of LRU address this?

**Comment by dang** [comment]
There's actually been LRU / readdir fixes since 2.7.6, so yes, I believe so.

**Comment by ffilz** [comment]
Closing as resolved
